Migrate from ActiveMQ to Stockpile

senior opened the pull request:
This PR merges the cmd-proc-refactor branch into master and moves from ActiveMQ to Stockpile. Some parts of this PR have been reviewed in prior PRs (before going into cmd-proc-refactor) and others have not. The code that has been reviewed, often had a caveat of, "this is good enough for now" or "this will change in the next set of commits" etc. So this review is intended to cover those situations.

In particular, I'm looking for review comments to be in one of the three categories.

1. Things that are obviously bad/wrong/inconsistent and that need to be fixed/reworked/cleaned up before we merge into master.

2. Significant changes/refactoring that could/should be done soon but after we merge to master (before we cut a release). An example of this is merging the mq_listener and command namespaces (ticketed at [PDB-3038](https://tickets.puppetlabs.com/browse/PDB-3038)).

3. Significant changes/refactoring that are enabled by this change, but not critical to get in before we release. We'll ticket these and put them in the backlog.

Below are the changes we know need to be made before releasing, but were planning to work on after merging to master

- We currently store commands with a filename <integer>_<received_time>_<command>_<version>_<certname>.json where command is our normal command names (i.e. "replace catalog"). We should have something more concise and without spaces [PDB-2982](https://tickets.puppetlabs.com/browse/PDB-2982)

- Deprecate old ActiveMQ and DLO related config that is no longer relevant [PDB-2948](https://tickets.puppetlabs.com/browse/PDB-2948)

- Ensure the new stockpile code is HUPable [PDB-2983](https://tickets.puppetlabs.com/browse/PDB-2983)

- Ensure that in the unlikely case that we have a hugely long certname, that we gracefully handle filenames that are too long [PDB-3039](https://tickets.puppetlabs.com/browse/PDB-3039)

- Merge the mq_listener and command namespaces [PDB-3038](https://tickets.puppetlabs.com/browse/PDB-3038)

There are several performance enhancements we're planning to make before a release, but those should be independent of this set of core functionality.

puppetlabs-jenkins commented:
Test FAILed

senior commented on project.clj:
```diff
@@ -36,6 +36,7 @@
   :dependencies [[org.clojure/clojure "1.8.0"]
                  [puppetlabs/i18n ~i18n-version]
                  [cheshire "5.6.1"]
+                 [digest "1.4.4"]
```
Adding this dependency to hash bytes, other functions we have (that use this same library) hash strings. We have a difference in version here and in kitchensink, we should probably switch to 1.4.3 so we don't conflict

senior commented on resources/public/dashboard/index.html:
```diff
@@ -257,11 +257,6 @@
        url: "/metrics/v1/mbeans/puppetlabs.puppetdb.storage:name=gc-time",
        format: d3.format(",.3s"),
        snag: getPercentileSnag},
-      {description: "DLO Compression",
```
I think @rbrw can comment more on this, but we found that DLO as it is written, wasn't "connected up", so it was impossible to hit that code, other than in tests. We made the decision that if we've gone this far without it, no need to reimplement it with the new DLO code.

wkalt replied:
that's correct, this code was dead

senior commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -361,29 +328,31 @@
+    ;; Error handling here?
+    (let [stockdir (conf/stockpile-dir config)
+          command-chan (async/chan
+                         (if (:catalog-facts-bash command-processing)
```
This config option enables "bash in place" where catalogs/factsets for a given certname are marked as "delete" when a new catalog/factset from the same certname is enqueued. The new catalog/fact is put at the tail of the queue and won't be processed sooner, but the old catalogs/factsets won't be process, just deleted from disk when they are at the head of the queue

rbrw replied:
And we only mark the commands as deleted so that we can avoid doing anything involving the messy filesystem world while holding the queue locks.  The deletion is handled (more easily) by the thread that eventually pulls the deleted command off the queue.

senior commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -361,29 +328,31 @@
+          command-chan (async/chan
+                         (if (:catalog-facts-bash command-processing)
+                           (queue/sorted-command-buffer
+                             max-enqueued
```
Max enqueued is the maximum number of command references that we can have in memory. There's more on command refs below, but when we reach that number (i.e. 1 million) we'll block on put to the channel, thus holding up enqueues until one message has been processed.

senior commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -361,29 +328,31 @@
+          clean-lock (ReentrantLock.)
+          command-loader (when load-messages
+                           (future
+                             (load-messages command-chan mql/inc-cmd-metrics)))]
```
This call enqueues all existing stockpile messages into the command channel

senior commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+(ns puppetlabs.puppetdb.cli.shovel
```
This namespace handles migrating from an "old" ActiveMQ version of PuppetDB to stockpile. My thinking here is taht we should be making a "best effort" to migrate existing messages. If we have some failure, or some messages can't be read (i.e. KahaDB corruption maybe) I don't think we should fail to start. Would be interested to hear if folks feel differently.

rbrw replied:
Note too that if we get to the end of the migration, we'll mark the queue as "migrated" and never try again.

rbrw replied:
Might also be worth noting that this code also "upgrades" all commands to the current POST-with-headers style, so that now, all other PDB code can assume it will only see the latest format.

wkalt replied:
any feelings on changing the name of this ns? I understand why it's called shovel but anyone new to the code is going to find it weird. puppetlabs.puppetdb.cli.queue-migration or something else explicit maybe?

senior replied:
I'd probably put activemq or amq in the name, maybe `puppetlabs.puppetdb.cli.amq-migration`?

wkalt replied:
works for me

senior commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -450,9 +443,11 @@
           response-mult (async/mult response-chan)
-          response-chan-for-pub (async/chan)]
+          response-chan-for-pub (async/chan)
+          concurrent-writes (get-in (get-config) [:command-processing :concurrent-writes])]
```
I'm thinking this will go away once we merge into master. There were some questions around how many concurrent writes we should allow, and what would be the optimal value. @wkalt's research has found it doesn't have a huge impact on performance. My thinking now is that we don't constrain writes and just allow the jetty threads to write without issue

wkalt replied:
+1

senior replied:
The opposite of this is true (noted in other comments). This is very important for performance with spinning disks. This code will be staying.

senior commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -461,60 +456,44 @@
-             :factory factory
-             :connection connection)))
+
+      (upgrade-activemq (get-config)
```
Currently the ActiveMQ upgrade occurs in the foreground. We've been giving some thought to this happening in a future.

wkalt replied:
foreground seems reasonable to me. I think in 99% of cases the queue will be either small or empty. Putting it in a future adds a bit of complexity and may not buy us much.

senior commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -428,20 +414,27 @@
       (throw ex))))
 
+(defn upgrade-activemq [config enqueue-fn dlo]
+  (when (shovel/needs-upgrade? config)
```
Standing up the ActiveMQ stack takes quite a while, we should only do that if we know we need to

senior commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -428,20 +414,27 @@
                   (fn []
                     (process-command! cmd db)))]
       (swap! stats-atom update :executed-commands inc)
+      (callback {:command cmd :result result})
```
This callback is the one attached to the command ref, for use on our "blocking post" feature

senior commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+      {:info info-dest
+       :command cmd-dest})))
+
+(defn discard-bytes
```
The only real purpose of this function is to discard a command when we don't know much about it. Specifically this happens when migrating an ActiveMQ message into stockpile and some error occurs. If it's an old message and we can't parse the payload, we won't know the command type, certname, etc. This allows us to put that message somewhere that the user can follow up on

senior commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+                                               attempts
+                                               path)]
+      (swap! metrics ensure-cmd-metrics registry command)
+      (update-metrics @metrics command (Files/size cmd-dest))
```
Note that we are updating DLO metrics as they are discarded. We only read from the filesystem on startup, from there it's only updated in memory.

kbarber commented:
@puppetlabs-jenkins retest this please

senior commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -84,31 +84,30 @@
+  [do-submit-fn uuid completion-timeout-ms]
+  (let [command-result (promise)]
+    (do-submit-fn (fn [the-result]
+                    (deliver command-result the-result)))
```
Previously this would watch a core.async/channel for results, this is now a callback with a promise. Note that this (like before) doesn't persist past restarts.

senior commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -168,32 +167,38 @@
       (instance? java.io.InputStream body)
-      (IOUtils/toByteArray
-       (restrained-drained-stream body (long max-command-size)))
+      body
```
Yay! You can stream commands to the queue!

senior commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -168,32 +167,38 @@
+      (throw (Exception. (str "Unexpected body type: " (class body)))))
+    (cond
+      (instance? java.io.InputStream body)
+      (restrained-drained-stream body (long max-command-size))
```
It seems like we should revisit this behavior at some point. At the time, it was a memory issue. Now it's not an issue of enqueuing, but rather an issue of dequeuing. We could stream it to disk, then move it to the DLO if it was too large. This also opens up a potential DoS, what if people throw lots of large messages our way and fill up the disk. Something to think about here.

rbrw replied:
Good point(s), though I suspect we may have any number of other places we'll need to address if we want to be more careful on that front (e.g. if postgresql is on the same disk, then you could just feed it "too much" actual data?).

senior commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -207,16 +212,21 @@
          (do
-           (do-submit)
-           (http/json-response {:uuid uuid}))))
+           (do-submit identity)
```
Currently all command refs include a callback. They will all be identity except in the blocking post case.

senior commented on src/puppetlabs/puppetdb/import.clj:
```diff
@@ -36,17 +60,32 @@
-                      (utils/read-json-content tar-reader)))
+                      (extract-certname "catalogs" path)
+                      (-> tar-reader
+                          utils/read-json-content
```
This is a hack around the awkward TAR lib we're using. It gives 1 input stream that is "marked" for entries in the tar. What we really want here is a "substream" kind of thing where we can pass Stockpile a stream that is a slice into that big stream, i.e. byte 1,000 through 75,000. AFAIK, that's not a thing and if it was, we'd still need to make some significant changes here to make it work. I just hacked around it with a pretty inefficient call to generate-string and converted it a ByteArrayInputStream

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
-               json/generate-string
-               (publish-fn delay)))))
-
-(defn create-message-handler
```
I've collapsed down most of this middleware. I think there is some simplification that can be done beyond this, but my initial goal was being able to see all of the pieces together. I think with stockpile, this code is much easier to test and breaking it up the way it was before is no longer necessary.

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+     command-result)))
+
+;; Primarily a test hook
+(defn discard-message [message q dlo]
```
I remember talking with @rbrw was it the existing code that required this?

rbrw replied:
No, we still use this insinuation-point in one of the tests; we might be able to do without it, but seems harmless, and at the time, the alternatives weren't appealing.

senior replied:
what's the downside of using dlo/discard-cmdref as the test hook?

rbrw replied:
I suspect that may be fine at this point; vaguely recall that earlier, there might have been more going on.

senior commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -368,12 +363,13 @@
-    "store report" (store-report command db)
-    "deactivate node" (deactivate-node command db)))
+  [{command-name :command version :version delete? :delete? :as command} db]
+  (when-not delete?
```
When the delete? metadata is found, we skip processing

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+     ;; below, it's only going to be removed
+     (if delete?
+       (do
+         (process-message cmdref)
```
I don't recall now why we're processing this message on a delete. When it gets further down the stack, it just gets ignored. Maybe it'd be ok to avoid this call entirely?

rbrw replied:
Sounds good, as long as no one depends on the idea that the process-message callback will be called for every message (of course).

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+     (if delete?
+       (do
+         (process-message cmdref)
+         (queue/ack-command q {:entry (queue/cmdref->entry cmdref)})
```
This will remove the command from disk

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+    (try+
+     ;; If the message is a delete?, there's no need to parse it
+     ;; below, it's only going to be removed
+     (if delete?
```
This is a bit ugly, but we can to avoid parsing the message if we can. When the delete? flag is present we can avoid that, delete it and move on

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+(defn discard-message [message q dlo]
+  (dlo/discard-cmdref message q dlo))
+
+(defn message-handler
```
This function is a bit long, but it's all of the middleware that was in the codebase previously, collapsed down to a single function.

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+                  (log/debug ex log-str)
+                  (-> cmd (queue/cons-attempt ex) delay-message))
+
+                (< retries maximum-allowable-retries)
```
This fixes a bug currently in master. In master, the last attempt is skipped. This patch will try up to `maximum-allowable-retries` then fail (else clause below)

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -330,64 +278,76 @@
+(defn send-delayed-message [command-chan delay-pool]
+  (fn [cmd]
+    (let [narrowed-entry (dissoc cmd :payload)]
+      (after one-hour #(async/>!! command-chan narrowed-entry) delay-pool))))
```
Previously failed commands were enqueued to the scheduler queue, that was periodically polled, then re-enqueued in the regular queue. This behavior is now handled in memory. `delay-pool` is a Java scheduled threadpool, managed by atat. Failures require no disk writes. This function removes the payload to consume less memory. That command ref stays in the delay-pool for one hour, then invokes the above function, which re-enqueues the command ref.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (when-let [^CommandRef old-command (.get certnames-map [command-type certname])]
```
What happens when an older message is re-enqueued (i.e. after a 1 hour retry). I think as-is, this will overwrite the new command with the old command.

senior commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -330,64 +278,76 @@
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size]
+  (gtp/create-threadpool size "cmd-proc-thread-%d" 10000))
```
This `10000` is an arbitrary number that could probably use some more thought. This is the number of milliseconds we will wait after telling the threadpool to shutdown before we start forcibly shutting down things.

wkalt replied:
What's the consequence of a forcible shutdown here? A couple lost messages?

senior replied:
No lost messages. Depending on the timing it could mean abandoning a transaction in the middle. It could mean the transaction was committed, but we haven't removed the message from disk yet (we'd reprocess it on restart) etc.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+(ns puppetlabs.puppetdb.queue
```
This ns is PuppetDB's interface to stockpile

rbrw replied:
Further, it's where all the commands "in flight" live, i.e. for every command that PDB knows about there'll be a cmdref somewhere in here, backed by a stockpile entry referring to a file on disk, which then allows pdb to have a "global" view of what's pending.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+(def metadata-command-names
+  (vals constants/command-names))
+
+(defn stream->json [^InputStream stream]
```
All of these functions could use docstrings

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+(defn stream->json [^InputStream stream]
+  (try
+    (-> stream
+        (InputStreamReader. StandardCharsets/UTF_8)
```
This StandardCharsets/UTF_8 should be a constant somewhere, right now we have a mixture of this and the "UTF-8" strings

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+  (format "%d_%s_%d_%s.json"
+          (tcoerce/to-long received) command version certname))
+
+(defn- metadata-rx [valid-commands]
```
Not sure why this is private

rbrw replied:
Heh, well I'm sure you know who did it; I probably did that because if felt like internal machinery, but it's certainly not critical.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+        (when-let [[_ id qmeta] (re-matches rx s)]
+          (parse-metadata qmeta))))))
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
```
This is the "ticket stub" that PuppetDB uses to dequeue messages. This needs to be as lean as possible as there could be many of them in memory. When we switch from the proper command names to a short hand when writing the filenames to disk, we should be able to intern command, version is a number along with received.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+             :entry entry))))
+
+(defn cons-attempt [cmdref exception]
+  (update cmdref :attempts conj {:exception exception
```
@rbrw shouldn't attempts be added to the defrecord?

rbrw replied:
We certainly could; whether that's a win or a loss I suppose depends on what fraction of commands have failures before being processed.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
+
+(defn cmdref->entry [{:keys [id command version certname received]}]
```
Entry is was stockpile uses to refer to a message, basically an id and a metadata string. Rather than store this string in memory, we generate it when needed.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+                      :annotations {:id (stock/entry-id entry)
+                                    :received received}})))
+
+(defn cmdref->cmd [q cmdref]
```
This is a hydrated command with a parsed payload

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+  (count [this]
+    (.size fifo-queue)))
+
+(defn sorted-command-buffer
```
This buffer is what is being used to bend the properties of the queue a bit. It's using a java `TreeMap` to keep a FIFO order of the messages based on the stockpile entry-id (a Long). The second HashMap keeps track of the the most recent catalog/factset for a given certname. This `HashMap` is consulted when a new catalog/factset comes in so that we can mark the existing one as deleted.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+   ;; accepting a function here is a hack to get around a cyclic dependency
+   ;; between this ns, mq-listener.clj, and dlo.clj. My hope is we'll be able
+   ;; to get rid of it somehow when we refactor mq-listener and command.clj.
+   (SortedCommandBuffer. (TreeMap.) (HashMap.) n delete-update-fn)))
```
This `delete-update-fn` is our hook to get metrics on the deleted catalogs/factsets. Ideally this would go in command.clj or mq_listener.clj, but @wkalt ran into some dependency issues. My hope is that this goes away once we merge together those two namespaces.

senior commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+                    chan)
+                  command-chan)))
+
+(defn create-or-open-stockpile
```
There's two pieces to opening a stockpile instance. The opening of the queue itself so that you can enqueue more stuff, then there's what to do with the existing messages. This returns the Stockpile queue instance and a function that can be called when ready to read in the existing messages from the queue. This allows enqueuing of old messages will starting up (and/or received new messages).

senior commented on test/puppetlabs/puppetdb/admin_clean_test.clj:
```diff
@@ -66,17 +66,22 @@
           orig-clean cli-svc/clean-up
           in-clean (CyclicBarrier. 2)
-          test-finished (CyclicBarrier. 2)]
+          test-finished (CyclicBarrier. 2)
```
The changes in this namespace aren't necessarily stockpile/activemq related, but @rbrw and I fixed a race condition and knowing that this PR was coming, put the commits here instead.

senior commented on test/puppetlabs/puppetdb/cli/services_test.clj:
```diff
@@ -65,13 +66,19 @@
-                                 :bar "the bar"
-                                 :baz "the baz"}
-                        :producer_timestamp (to-string (now))})
+      (enqueue-command dispatcher
```
This code is stale, I have some test utility code that accepts a clojure data structure and coerces to a stream, switching to that would simplify this code and the test below it

senior commented on test/puppetlabs/puppetdb/cli/shovel_test.clj:
```diff
@@ -0,0 +1,141 @@
+    (mapv (comp str #(.toAbsolutePath %))
+          (-> path-stream .iterator iterator-seq))))
+
+(deftest shovel-upgrade-test
```
Do we need more tests here? I'm testing a happy path upgrade, an upgrade with a "bad message" and a "no upgrade" scenario.

puppetlabs-jenkins commented:
Test FAILed

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -96,37 +104,62 @@
-       (finally
-         (fs/delete-dir discard-dir#)))))
+  [command delay-var discard-var db & body]
+  `(tqueue/with-stockpile q#
```
I tried to keep the changes to this test code to a minimum. I think that it's simpler to test these pieces integrated now, it might not be as necessary to have all the moving parts. The existing tests work though, just switching the calls like I have here

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1276,11 +1325,11 @@
                         :payload {:certname "bar.example.com"
                                   :producer_timestamp (now)}}}
-             {:certname "bar.example.com"
+             #_{:certname "bar.example.com"
```
I'll revisit these two tests. I remember something about them not being valid anymore. I'll either confirm that for fix it.

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1505,10 +1554,15 @@
+          (enqueue-command (command-names :replace-facts)
+                           4
+                           "foo.local"
+                           (-> {:environment "DEV" :certname "foo.local"
```
This should be updated to the test store-command function, it'll coerce to a stream automatically

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1526,8 +1580,13 @@
           expected-stamp (DateTime. deactivate-ms DateTimeZone/UTC)]
-      (enqueue-command (command-names :deactivate-node) 3
-                       {:certname "foo.local" :producer_timestamp input-stamp})
+      (enqueue-command (command-names :deactivate-node)
```
This should use the test coerce to stream function

senior commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1557,15 +1616,112 @@
+      (enqueue-command (command-names :deactivate-node)
+                       3
+                       "foo.local"
+                       (-> {:certname "foo.local" :producer_timestamp producer-ts}
```
This should use the test coerce to stream function

senior commented on test/puppetlabs/puppetdb/status_test.clj:
```diff
@@ -19,6 +19,9 @@
         (is (= {:maintenance_mode? false
                 :read_db_up? true
                 :write_db_up? true
+                ;; Stockpile needs metrics implemented around it to
```
This can be removed now

senior commented on test/puppetlabs/puppetdb/status_test.clj:
```diff
@@ -35,5 +38,8 @@
           (is (= {:maintenance_mode? true
                   :read_db_up? true
                   :write_db_up? true
+                  ;; Stockpile needs metrics implemented around it to
```
This can be removed now

senior commented on test/puppetlabs/puppetdb/testutils.clj:
```diff
@@ -26,6 +27,17 @@
             [clojure.set :refer [difference]]
             [puppetlabs.puppetdb.test-protocols :as test-protos]))
 
+(defn ordered-matches?
```
@rbrw is this still used?

rbrw replied:
Yeah, we are, but I think I still have a version without it (using the big regex) if we want to switch, or we could decide we just don't care about checking the err file as much, and do something even simpler if we like.

senior replied:
Do we need to have an extensions PR that removes that from extensions and uses this one?

rbrw replied:
Sure, if/when it ends up in master; this version's a bit nicer too.

rbrw commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -257,43 +247,22 @@
-    (when exists?
-      (mq/remove-queue! "localhost" "com.puppetlabs.puppetdb.commands")
-      (log/info "Removed legacy queue"))))
-
```
Note: these should be handled by the mq -> stockpile shovel now, below.

rbrw commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
 (ns puppetlabs.puppetdb.command.dlo
-  (:import [org.apache.commons.io FileUtils])
```
The DLO changes substantially here - discarded messages mirror the stockpile files they come from, and for each foo.json discarded command, there is a foo-err.txt detailing the full backtrace for each attempt.

Support for archiving, and per-command subdirs is gone.

rbrw commented on src/puppetlabs/puppetdb/metrics/core.clj:
```diff
@@ -2,10 +2,12 @@
             [metrics.core :refer [new-registry]]))
 
-(defn new-metrics [domain]
+(defn new-metrics [domain & {:keys [jmx?] :or {jmx? true}}]
```
This makes it possible to write tests that use metrics, but don't fire up (possibly multi-instance or parallel-test conflicting) jmx support.

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+  [^MessageConsumer consumer process-message id]
+  (try
+    (loop []
+      (when-let [msg (.receive consumer 5000)]
```
we should probably make sure we test this with a super large queue on startup. Anecdotally speaking some of the AMQ components can take a long time to get started/moving, but I don't have a good sense of where the holdups occur or whether they'd affect this.

senior replied:
Not sure if this helps the understanding or not, but startup happens in serial, we won't start trying to receive a message until the broker has started, connection factor has been initialized etc.

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+            [puppetlabs.trapperkeeper.logging :as logutils]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [puppetlabs.puppetdb.config :as conf]
+            [puppetlabs.puppetdb.utils :as utils]
```
utils, conf, kitchensink, logutils, and config are unused

wkalt replied:
I'd say just generally check this for unused stuff. I haven't checked the imports.

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+  "Converts from the command format that didn't include the certname
+  in the toplevel key list and had all the command data with the
+  payload key"
+  [headers message-bytes]
```
this doesn't use the headers

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+  "Commands coming out of this function are suitable for enqueuing
+  into stockpile. They must include command, version, certname and a
+  payload"
+  [{:keys [command version certname] :as headers} message-bytes]
```
no need to destructure as headers here since convert-old-command-format doesn't use them

senior replied:
I don't think I agree. `convert-old-command-format` doesn't use it, but it's used twice in the function. Are you suggesting I pull those values out twice by keyword instead of once by destructuring?

wkalt replied:
I'm saying it could just be
```
[{:keys [command version certname]} message-bytes]
```

headers isn't used in this function.

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+                                             (catch Exception e
+                                               (discard-amq-message (convert-message-body msg) (swap! id inc) {} e dlo))))]
+          (.send producer request)
+          (Thread/sleep 2000)
```
this seems odd

senior replied:
@ajroetker why is this here and why 2 seconds?

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+                                               (discard-amq-message (convert-message-body msg) (swap! id inc) {} e dlo))))]
+          (.send producer request)
+          (Thread/sleep 2000)
+          (log/info (i18n/trs "Processing messages from the Scheduler"))
```
scheduler should be lower case

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+          (consume-everything consumer process-message-and-remove id))))))
+
+(def default-mq-endpoint "puppetlabs.puppetdb.commands")
+(def retired-mq-endpoint "com.puppetlabs.puppetdb.commands")
```
is this to support upgrades from 2.x or something? I thought the com stuff went away a long time ago.

ajroetker replied:
correct, we were still supporting upgrades from 2.x when I wrote this, not sure if that's changed

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+  (try
+    (start-broker! (build-embedded-broker brokername dir config))
+    (catch java.io.EOFException e
+      (log/error e (i18n/trs "EOF Exception caught during broker start, this might be due to KahaDB corruption. Consult the PuppetDB troubleshooting guide."))
```
I know this code has just been moved, but the comma should be a period.

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+  [config]
+  (fs/touch (upgrade-lockfile-path (get-in config [:global :vardir]))))
+
+;;
```
this can be removed

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+              (log/info (i18n/trs "You may safely delete {0}" mq-dir))
+
+              (catch Exception e
+                (log/error e (i18n/trs "Uable to receive ActiveMQ messages. Migration of existing messages failed.")))
```
Unable

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+              (finally
+                (stop-broker! broker))))
+          (catch Exception e
+            (log/error e (i18n/trs "Uable to start ActiveMQ broker. Migration of existing messages failed.")))
```
Unable

wkalt commented on src/puppetlabs/puppetdb/cli/shovel.clj:
```diff
@@ -0,0 +1,374 @@
+          (finally
+            (.stop conn-pool))))
+      (catch Exception e
+        (log/error e (i18n/trs "Unable to connect to ActiveMQ. Migration of existing messages failed."))))))
```
👍

wkalt commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -270,26 +266,25 @@
-  [{:keys [payload annotations version]} db]
-  (let [{:keys [id]} annotations
-        {:keys [certname values] :as fact-data} payload
+  [{:keys [payload annotations version] :as command} db]
```
command is not used

ajroetker replied:
neither is version it appears

senior replied:
I think version should be gone. I think command is useful as it says the kind of map that is passed in, as opposed to a map that has a payload an annotations. If it's more distracting than useful though, I can remove it.

wkalt commented on src/puppetlabs/puppetdb/config.clj:
```diff
@@ -149,7 +149,15 @@
+     ;; or expected to be released. Perf testing will confirm if the
+     ;; parameters are needed and what their ideal value would be in our
+     ;; various deployment scenarios
+     :concurrent-writes (pls/defaulted-maybe s/Int 100)
```
I guess we'll be removing these from here too -- this is the PR to master, so is now the time to do that?

senior replied:
I was thinking I'd remove it after the merge to master, but I can remove it now

senior replied:
Update on this, we're not removing it, it turns out to be pretty important, I created ticket PDB-3060 to make this config param "official"

wkalt replied:
we can remove the comment above

senior replied:
@wkalt are you thinking we'd want to let users tune the max-enqueued as well?

wkalt commented on src/puppetlabs/puppetdb/config.clj:
```diff
@@ -149,7 +149,15 @@
+     ;; various deployment scenarios
+     :concurrent-writes (pls/defaulted-maybe s/Int 100)
+     :max-enqueued (pls/defaulted-maybe s/Int 1000000)
+     :catalog-facts-bash (pls/defaulted-maybe String "false")}))
```
likewise this -- it'll never make sense to have it false

wkalt commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -84,31 +84,30 @@
+                             :processed false
+                             :timed_out true}
+                            http/status-unavailable)
+        (let [{:keys [command exception]} result
```
command is unused

wkalt commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -168,32 +167,38 @@
       ;; Given Java's UCS-2 encoding, incoming UTF-8 more or less
       ;; doubles in size when converted to a String.
-      (if (> (count body) (* 2  max-command-size))
+      (if (> (* 2 (count body)) max-command-size)
```
nice catch

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -1,21 +1,30 @@
             [metrics.histograms :refer [histogram update!]]
             [metrics.timers :refer [timer time!]]
-            [puppetlabs.puppetdb.command :as cmd]
+            [metrics.counters :refer [counter inc! dec! value clear!]]
```
we can get rid of clear!

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -330,64 +278,76 @@
   MessageListenerService
   [[:DefaultedConfig get-config]
-   [:PuppetDBServer]] ; MessageListenerService depends on the broker
+   [:PuppetDBServer shared-globals]] ; MessageListenerService depends on the broker
```
comment can go

senior commented on src/puppetlabs/puppetdb/config.clj:
```diff
@@ -149,7 +149,15 @@
+     ;; parameters are needed and what their ideal value would be in our
+     ;; various deployment scenarios
+     :concurrent-writes (pls/defaulted-maybe s/Int 100)
+     :max-enqueued (pls/defaulted-maybe s/Int 1000000)
```
What should we do about this one? I would rather not have a config item for it, but it seems like something we probably should.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

kbarber commented:
Going from 3.0.0 -> this PR results in MQ exceptions during migrations: https://gist.github.com/kbarber/5f3e42830ace3fe54c778c576cdfab51 ... you can repeat this by running 3.0.0, stopping db, run a puppet agent -t to submit 3 commands, switch to this patch (with db up) and start the service.

kbarber commented:
@senior looks like this needs a rebase.

senior commented:
@kbarber I fixed the issues you found and rebased. There were two separate issues. The first was the code assumed that version was a string. This is the case on new messages (it comes from an HTTP parameter) but not the case for older messages where the version is in the body of the request (it's already an integer). I removed that code and updated the test.

The second problem was just a misleading exception. It stemmed from closing the AMQ connection pool after closing the broker. I reversed the order of the close and it made the second exception go away.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

rbrw commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -214,50 +223,37 @@
+    (finally
+      (.release write-semaphore)
+      (when command-stream
+        (.close command-stream)))))
```
I can't recall if we already discussed this, but naively, I might be inclined to have the caller manage the closing of the stream if that's feasible (i.e. in an easy to inspect with-open or whatever); if not, then let's mention the close in the doc string.

senior replied:
The caller in this case is Jetty, so it's already opened when we get it, I can add it to the docstring

rbrw commented on src/puppetlabs/puppetdb/amq_migration.clj:
```diff
@@ -0,0 +1,369 @@
+              ;; we're processing
+              (create-scheduler-receiver conn-pool mq-endpoint process-message dlo id)
+              (create-mq-receiver conn-pool mq-endpoint process-message id)
+              (create-mq-receiver conn-pool retired-mq-endpoint process-message id)
```
Not critical, but if we have another rev, I might consider changing these names to transfer-scheduled-messages and transfer-endpoint-messages or something...

rbrw commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -361,29 +328,29 @@
+                                                [:dlo :registry]))
+                   :command-chan command-chan}
+          clean-lock (ReentrantLock.)
+          command-loader (when load-messages
```
Not sure this is an issue, but I wondered about what happens if any of the remaining code in this function throws an exception after we start the future, i.e. does the tk lifecycle say that stop will be called, or do we know that the loader will behave reasonably regardless?

senior replied:
I'm not sure what you're asking here. If the stop isn't called, I'm assuming it'd exit? In that case I think we're good. If it does call stop then we cancel the future. Is there something I'm missing here?

rbrw replied:
I was wondering if say we launch this future, and then say maybe-check-for-updates throws an exception; will the future just keep running?  If so, do we care?  (And if we do, we might need a try/... cancel, that we can skip if we actually make it to the end of start or whatever...)

rbrw commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -428,20 +414,27 @@
         (recur (dec n))
```
I assume the Throwable catch above is what pdb already does, but while reviewing this I wondered about it, i.e. do we really want to keep trying after a virtual machine InternalError, etc.

senior replied:
My guess is where you see catch Throwable, the purpose it served was catching AssertionErrors. In other areas I've been splitting that out, my guess is that's the same reason it was originally used here.

rbrw commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -428,20 +414,27 @@
+(defn upgrade-activemq [config enqueue-fn dlo]
+  (when (mig/needs-upgrade? config)
+    (mig/activemq->stockpile config enqueue-fn dlo)
+    (mig/lock-upgrade config)))
```
I suppose right now, this means that if some exception that makes it back up to here, we'll retry the upgrade again next time; might be fine (and if that makes it possible for a user to get stuck in a cycle, suppose we can always tell them to touch the file themselves) -- just noting.

senior replied:
Or remove the AMQ dir

wkalt commented on src/puppetlabs/puppetdb/amq_migration.clj:
```diff
@@ -0,0 +1,369 @@
+  [^MessageConsumer consumer process-message id]
+  (try
+    (loop []
+      (when-let [msg (.receive consumer 5000)]
```
it looks to me like this returns nil if it times out rather than an exception -- not clear to me that we're handling that case right (seems like things just fall through silently and we incorrectly recommend queue deletion).

senior replied:
You're correct in your assessment here and it was intentional. Background here is the messages are already in the queue, we have properly started the broker, we are not allowing more messages to be enqueued. If there are no messages available, we must be done. Knowing we are done is the tricky part. The above code assumes that if we can't get a message in 5 seconds, we must be done. The time was totally arbitrary. From what I've read, the only way I have seen that is different, is to use the JMX metrics (which we know to be slightly out of date). It would be possible to combine the approaches. I'm open to that but I don't know that it would give me much more confidence in the correctness here. We can talk about some of the other things I've tried if you're interested @wkalt.

wkalt commented on src/puppetlabs/puppetdb/amq_migration.clj:
```diff
@@ -0,0 +1,369 @@
+                                             (catch Exception e
+                                               (discard-amq-message (convert-message-body msg) (swap! id inc) {} e dlo))))]
+          (.send producer request)
+          (Thread/sleep 2000)
```
this is still here, can't remember the story

senior replied:
I think the comment got wiped, I asked @ajroetker and didn't hear back, trying a second ping.

ajroetker replied:
I stole this code from the AMQ tests and they had a comment about this being a safety measure to make sure the request had been processed by the producer before looking for the messages from the scheduler

wkalt commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+    {:messages (counters/counter registry [dns category "messages"])
+     :filesize (counters/counter registry [dns category "filesize"])}))
+
+(defn- update-metrics
```
it would be nice if the metrics-related functions in here had names that clearly connected them to the DLO

rbrw replied:
Hmm, well I'd done this under the assumption that the namespace would be aliased, so you'd have dlo/initialize, etc;  i.e. if we're going to prefer using prefixes (:as), then otherwise you'd have dlo/dlo-initialize, which seems a little redundant.

rbrw replied:
Though plausible counter-argument to myself -- if we're not going to use consistent aliases, then your suggestion is strictly better since it means that things like `grep dlo-initialize` won't be half-useless (e.g. if it's not always dlo/initialize).

wkalt commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+
+(defn- entry-cmd-data-filename
+  [entry]
+  ;; 10291-1469469689-cat-4-foo.org.json
```
I'd rather these be docstrings with another explanatory word or two.

wkalt replied:
several instances in this file

wkalt commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+  (let [tmp (Files/createTempFile dir
+                                  (str "tmp-err-" id \- command) ""
+                                  (make-array FileAttribute 0))]
+    ;; Leave the temp file if something goes wrong.
```
What's the reason for the temp file exactly? i.e why not just write to the final destination?

wkalt replied:
and can we stick the answer in the doc string

senior replied:
The reason is atomicity. What happens if you kill the process while halfway through writing the bytes to the file? With the current approach, the bytes are written, then it's atomically renamed to the correct name. With this you're always sure that if you're looking at a properly named file, it must have all of the data. If it's a tmp- file, it's possible it was interrupted in the middle of writing it and might be mangled in some way.

rbrw replied:
Right, so you'll still have the data, but there's also to be likely to be an indication that "something happened", i.e. if the file is in place under the "final name", it's likely complete; though documenting that rationale here seems completely reasonable.

wkalt commented on src/puppetlabs/puppetdb/command/dlo.clj:
```diff
@@ -1,210 +1,182 @@
+        (Files/move tmp dest (copts [copt-replace])))
+      dest)))
+
+
```
extra line

wkalt commented on src/puppetlabs/puppetdb/config.clj:
```diff
@@ -158,6 +165,12 @@
    :reject-large-commands Boolean
+
+   ;;This is a test only config for now, see the related comment in
+   ;;command-processing-in
```
we can remove this

wkalt commented on src/puppetlabs/puppetdb/export.clj:
```diff
@@ -27,22 +27,25 @@
   [{:keys [certname producer_timestamp]}]
   (let [formatted-start-time (->> producer_timestamp
                                   time-coerce/to-date-time
                                   (time-fmt/unparse (time-fmt/formatter "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")))]
```
since we don't really care about the string formatting we can probably just call generic-identity-hash on producer_timestamp directly.

senior replied:
We don't care about the formatting, but it needs to be normalized so that it is always the same regardless of input format. Calling generic-identity-hash will perform the same conversion as above though, so I can switch it and avoid having that date formatting in two places.

wkalt commented on src/puppetlabs/puppetdb/export.clj:
```diff
@@ -27,22 +27,25 @@
   (let [file-suffix
         (case entity
-          "factsets" ["facts" (str (:certname datum) ".json")]
+          "factsets" ["facts" (str (:certname datum) file-suffix)]
```
I find the two meanings of file-suffix here pretty confusing

senior replied:
file-suffix is baked into our tar related code, I'll change the ".json" to file-extension to make the difference more clear

wkalt commented on src/puppetlabs/puppetdb/import.clj:
```diff
@@ -12,12 +12,36 @@
+  the characters after the certname"
+  (+ 1
+     40
+     (count export/file-suffix)))
```
realistically, how much value is there in putting file-suffix in a global variable? Are we ever going to do anything but .json for this?

senior replied:
Yes, we currently have plans to switch to json.gz before we release this code.

wkalt commented on src/puppetlabs/puppetdb/import.clj:
```diff
@@ -12,12 +12,36 @@
+  (let [prefix-length (+ export-root-dir-length
+                         (.length "facts")
+                         2)]
+    (subs path prefix-length (- (.length path) (count export/file-suffix)))))
```
seems like a lot of this code would be simpler if written in terms of regular expressions

senior replied:
Interesting, we just had a discussion around this, another vote in the favor of a regex here is I think we can use it here and instead of the file-pattern from line 18 above. I'll try switching it over and putting some tests around it.

senior replied:
I think this made the code much better, you can see the change [here](https://github.com/puppetlabs/puppetdb/pull/2061/commits/471684a1e41f6fe3348648da3b246d76d2179f5c)

wkalt commented on src/puppetlabs/puppetdb/import.clj:
```diff
@@ -36,17 +60,32 @@
-                      (utils/read-json-content tar-reader true)))
+                      (extract-certname "reports" path)
+                      (-> tar-reader
+                          (utils/read-json-content true)
```
we're keywordizing keys here but not in the other cases -- any reason for that?

senior replied:
Not sure, I'd like to keep the keys as string in general, but in this case I just left alone what was there.

wkalt commented:
this hits an NPE on a metrics update on the first command received to an empty queue

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+          (catch fatal? obj
+            (mark! (global-metric :fatal))
+            (let [ex (:cause obj)]
+              (log/error (:wrapper &throw-context) (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}" id command retries certname))
```
I'd break this

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+
+                :else
+                (do
+                  (log/error ex (i18n/trs "[{0}] [{1}] Exceeded max {2} attempts for {3}" id command retries certname))
```
line length

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,99 @@
+                      (discard-message q dlo)))))))))
+     (catch [:kind ::queue/parse-error] _
+       (mark! (global-metric :fatal))
+       (log/error (:wrapper &throw-context) (i18n/trs "Fatal error parsing command: {0}" (:id cmdref)))
```
length

wkalt replied:
79 for PEP-8 compliance.

wkalt commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+                       :callback command-callback
+                       :received (kitchensink/timestamp current-time)
+                       :annotations {:id (stock/entry-id entry)
+                                     :received (kitchensink/timestamp current-time)}}))))
```
given that we might have a lot of these in ram, could it be worth avoiding this duplication?

wkalt replied:
likewise we could save some bytes by representing command strings as integers

senior replied:
I'll write up a ticket for the id/received stuff. I kept it like this because everything in command.clj was looking in annotations. I think a top level key is better, so new code is looking there.

For command strings, this is being fixed in one of @aperiodic's PRs. We'll be using the same string for all commands of that type, so it'll only cost a reference for each command.

puppetcla commented:
Waiting for CLA signatures by @web-flow, @rbrw

@web-flow, @rbrw - We require a Contributor License Agreement (CLA) for people who contribute to Puppet, but we have an easy click-through license with instructions, which is available at https://cla.puppet.com/

Note: if your contribution is trivial and you think it may be exempt from the CLA, please post a short reply to this comment with details. http://docs.puppet.com/community/trivial_patch_exemption.html

wkalt commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -141,95 +174,111 @@
+  (testing "happy path, message acknowledgement when no failures occured"
+    (tqueue/with-stockpile q
+      (with-redefs [mql/discard-message (fn [& args] true)]
+        (let [mh (mql/message-handler q nil nil identity)
```
"handle-message"?

wkalt commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -141,95 +174,111 @@
+        (let [delay-message (mock-fn)
+              mh (mql/message-handler q nil delay-message
+                                      ;; Do we really want this to
+                                      ;; be ;; a RuntimeException?
```
we can probably remove this

wkalt commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -542,13 +581,11 @@
   (dotestseq [version catalog-versions
-              :let [command {:command (command-names :replace-catalog)
+              :let [command-1 {:command (command-names :replace-catalog)
                              :version latest-catalog-version
```
indentation

wkalt commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -594,14 +629,12 @@
+                               :version latest-catalog-version
+                               :payload basic-wire-catalog}
+                    command-2 (update-resource version command-1 "File" "/etc/foobar"
+                                               #(-> %(assoc :tags #{"file" "class" "foobar" "foo"})
```
weird spacing

wkalt commented on test/puppetlabs/puppetdb/command_test.clj:
```diff
@@ -1279,11 +1322,13 @@
-                        :payload (json/generate-string "bar.example.com")}}]]
+                        :payload (-> "bar.example.com"
+                                     json/generate-string
+                                     json/generate-string)}}]]
```
this looks crazy -- intentional?

senior replied:
Yeah. When an "old format" deactivate node command comes in, the certname **is** the payload. In http/command, we parse that payload, then generate-string it (that code is [here](https://github.com/puppetlabs/puppetdb/blob/master/src/puppetlabs/puppetdb/http/command.clj#L150)). In version 1 of deactivate node, our command processing code (not http/command, but puppetdb/command) expects a json escaped string. The outer "generate-string" gets unwrapped by http/command.clj and for version 1, puppetdb/command.clj is expecting json escaped there as well.

wkalt commented on test/puppetlabs/puppetdb/http/command_test.clj:
```diff
@@ -193,59 +230,78 @@
+            no-max-app (handler-with-max q command-chan false)
+            max-10-app (handler-with-max q command-chan 10)
+            req {:request-method
+                 :post :body "more than ten characters"
```
:post should be on the line above

wkalt commented on test/puppetlabs/puppetdb/queue_test.clj:
```diff
@@ -0,0 +1,198 @@
+           (parse-cmd-filename "0-0_replace catalog_0_foo.json")))
+    (is (= {:received r10 :version 10 :command "replace catalog" :certname "foo"}
+           (parse-cmd-filename "10-10_replace catalog_10_foo.json")))
+    (is (= {:received r10 :version 10 :command "unknown" :certname "foo"}
```
we should test a certname ending in .json

wkalt commented on test/puppetlabs/puppetdb/queue_test.clj:
```diff
@@ -0,0 +1,198 @@
+      (is (async/offer! c foo-cmd-2))
+      (is (= foo-cmd-2 (async/<!! c)))))
+
+  (testing "multiple older catalogs all get marged as deleted"
```
marked

wkalt commented:
If I submit a factset to puppetdb with this custom fact

```
Facter.add("mynullfact") do
  setcode do
    {"foo" => "\u0000"}
  end
end
```

master will bomb immediately but on this branch the message will stay in the queue with no logging, retries, or discards. This makes me think we've got some brokenness in the retries/DLO area that probably warrants a test of the full pipeline. This should include a test of this fact, and a test of a valid command during a simulated database outage, since those two cases behave differently on master (the invalid fact does not retry).

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -330,64 +278,76 @@
+        (try
+          (handle-message message)
+          (catch Exception ex
+            (log/error ex "Unable to process message. Message not acknowledged and will be retried")))))))
```
This is getting thrown in cases where the retry isn't happening -- right now I see it happening when an incoming report causes a schema violation due to a missing corrective_change field. I'm not sure if the issue is with logging or with retries not working.

wkalt commented:
The NPE I mentioned appears inconsistently -- probably a race condition.

puppetcla commented:
Waiting for CLA signature by @rbrw

@rbrw - We require a Contributor License Agreement (CLA) for people who contribute to Puppet, but we have an easy click-through license with instructions, which is available at https://cla.puppet.com/

Note: if your contribution is trivial and you think it may be exempt from the CLA, please post a short reply to this comment with details. http://docs.puppet.com/community/trivial_patch_exemption.html

puppetcla commented:
CLA signed by all contributors.

senior commented on test/puppetlabs/puppetdb/command/dlo_test.clj:
```diff
@@ -1,163 +1,205 @@
+         (is (= 1 (regval "unknown.messages")))
+         (is (= 1 (metval "unknown" :messages)))
+         (is (= unk-size (regval "unknown.filesize")))
+         (is (= unk-size (metval "unknown" :filesize))))))))
```
@rbrw is this something you can cover?

rbrw replied:
Should be.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

rbrw commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -428,20 +414,27 @@
                  (make-cmd-processed-message cmd nil))
       result)
     (catch Exception ex
+      (callback {:command cmd :exception ex})
```
I haven't figured out the full context yet, but just wanted to make sure we thought this was however bulletproof we need it to be, i.e. do we assume the callback doesn't throw (and is that made clear to the callback providers), or is it fine if it does?

senior replied:
from a practical perspective, this has two implementations, identity and [this](https://github.com/puppetlabs/puppetdb/pull/2061/files#diff-d82b3486b4cd22472cd542b7773086e9R90). I did hit an issue there in testing though (not supplying a callback when one was expected). We could surround this in a try catch and make it a best effort to be defensive though.

rbrw commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -84,31 +84,30 @@
+             http/status-unavailable)
+            (http/json-response (assoc base-response
+                                       :timed_out false)
+                                http/status-ok)))))))
```
May be just fine, but I wondered how the behavior of this might compare (in any way that matters) to what we had before if we shutdown or HUP with a number of items "in flight".

senior replied:
The behavior is broken in the face of restarts (both before and now).

rbrw commented on src/puppetlabs/puppetdb/import.clj:
```diff
@@ -4,19 +4,44 @@
+                                     (apply str
+                                            (map utils/regex-quote
+                                                 [utils/export-root-dir
+                                                  File/separatorChar
```
If I'm reading the broader context right, I wondered if we might not want th current platform separator here (i.e. if this is matching tar entries);  I'm not positive yet, but it's looking like the separator there might "always" be "/".

rbrw commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,104 @@
+               (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}" id command retries certname))
+              (-> cmd
+                  (queue/cons-attempt ex)
+                  (discard-message q dlo))))
```
If I'm reading this right, we have some cases in this function where we only update the metrics after the message is out of the queue, and others where we do the reverse; behavior should only differ if there's an unexpected exception, but I thought I'd mention it.

rbrw commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -133,174 +156,104 @@
+            (let [ex (:cause obj)]
+              (log/error
+               (:wrapper &throw-context)
+               (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}" id command retries certname))
```
Not sure we care, but looks like a logging exception (full disk?) will prevent the discard.

rbrw commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+  [q command]
+  (stock/discard q (:entry command)))
+
+(deftype SortedCommandBuffer [^TreeMap fifo-queue ^HashMap certnames-map ^long max-entries ^clojure.lang.IFn delete-update-fn]
```
Heh, ETOOLONG, even for github.

rbrw commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (.remove certnames-map [command-type (:certname cmdref)]))
+      cmdref))
```
Presumably it's either OK for this to return nil, or it won't ever be called for an empty fifo.

senior replied:
nils are never ok in core.async. If there's nothing there, it will block on take until something is AFAIK

rbrw commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+                (assoc old-command :delete? true))
+          (delete-update-fn (:command old-command) (:version old-command)))
+        (.put certnames-map [command-type certname] cmdref))
+      (.put fifo-queue (:id cmdref) cmdref))
```
It may not matter if any exception in here is more globally fatal to pdb, but if not then I wonder if we might want to reorder or something, i.e. would we want to make sure the new command is in the queue before we mark the old one deleted?

rbrw commented on src/puppetlabs/puppetdb/queue.clj:
```diff
@@ -0,0 +1,199 @@
+      (.put fifo-queue (:id cmdref) cmdref))
+    this)
+
+  (close-buf! [this])
```
Also not sure it matters for our purposes, but if there were a chance that someone might hold on to the buffer for a while after the close, wondered if .clear(ing) the maps might be friendly.  Even less extra work if we could just drop the map pointers, but that'd be uglier (mutable type?).

rbrw commented on src/puppetlabs/puppetdb/utils.clj:
```diff
@@ -290,6 +290,21 @@
+  [strings]
+  (format "(%s)" (->> strings
+                      (map regex-quote)
+                      (string/join "|"))))
```
Not sure, but wondered about having (?:%s), so that the caller can decide if they want a capturing group.

senior replied:
this is a commit I cherry-picked from @aperiodic I don't have an opinion, but if there was a change to be made, I'd vote for a separate commit that we add on in his PR.

rbrw commented:
ok, finished my current review round.

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test PASSed

puppetlabs-jenkins commented:
Test PASSed
